### ISSUE TITLE
bookmarks in query streams, with offset argument to websocket endpoints

### ISSUE THREAD

**Issue by S11001001** [digital-asset/daml#4509]
Starting at the end of non-live data (#4461) we will include a `"offset"` as discussed in #3878. This will incompatibly change the liveness marker from `{"live": true}` to the first occurrence of an offset: `{"events": [...], "offset": ...}`, where `[...]` *may* be empty.

As implied by the above, not every response block will include an offset. We only promise to provide offsets once we reach the end of non-live data. In fact, we promise *not* to provide offsets until we reach the end of non-live data.

The current query union format is `{templateIds, query} | [{templateIds, query}...]`, and fetch by key format is `[{templateId, key}]`; @hurryabit to determine where the nicest place to put `offset` is in these datatypes, noting that we can only reasonably have one offset for all queries in the union. The behavior of specifying a non-existent offset will be passed through from ledger API.

The only intended effect of this change is to make streams reconnect-able; the remaining stream semantics will not change.

This will disable "phantom archive" filtering entirely for resumed contract key streams, while it will remain in place for fetch streams for which offset is not supplied. Refining this behavior should be addressed in a separate issue, as @gerolf-da has noted various complicating factors.

Depends on #4465.

cc @dtanabe @lima-da 

* [ ] how to pass offset, especially in case of `[union...]`?
* [x] how to represent `LedgerBegin`? `null`

**Comment by S11001001** [comment]
@hurryabit Additionally we must have a JSON value for `LedgerBegin` in the model of `"offset"` in the stream:

https://github.com/digital-asset/daml/blob/master/ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala#L238-L254

This can occur when there is no non-live data, in which case we still want presence of the `"offset"` key to indicate you've reached live data. I guess `null` will be sufficient?

**Comment by hurryabit** [comment]
> This can occur when there is no non-live data, in which case we still want presence of the `"offset"` key to indicate you've reached live data. I guess `null` will be sufficient?

That sounds totally reasonable.



> @hurryabit to determine where the nicest place to put `offset` is in these datatypes, noting that we can only reasonably have one offset for all queries in the union.

I have no really good answer yet. What do you think about _preceeding_ the query message by an optional offset message in the formal `{offset: ...}`?

**Comment by leo-da** [comment]
@S11001001 how about 
```
{events: [...], live: false, offset=<last event offset>}
{events: [...], live: true,  offset=<last event offset>}
{events: [...], live: true,  offset=<last event offset>}
```

**Comment by S11001001** [comment]
@leo-da No, that won't work.

**Comment by S11001001** [comment]
> What do you think about _preceeding_ the query message by an optional offset message in the formal `{offset: ...}`?

This is kind of growing on me, since the query format doesn't change, and it doesn't imply anything that query can't do. @lima-da what do you think?

**Comment by lima-da** [comment]
@S11001001 
my first naive thought would be to wrap the union query in its own JSON object
```{"queries": [
    {"templateIds": ["Iou:Iou"], "query": {"amount": {"%lte": 50}}},
    {"templateIds": ["Iou:Iou"], "query": {"amount": {"%gt": 50}}},
    {"templateIds": ["Iou:Iou"]}
  ],
  "offset": ...
}```
the `offset` can be optional.
Compared to preceding an offset message before the actual query, this can start streaming immediately right after receiving the message, and we don't have to deal with timeout between offset and query, or separate error handle per message. 

This, of course, changes the query structure, but streaming endpoints is the only place allows union queries in payload, unless `POST /v1/query` will have the same union query, I don't see this approach is a query format change.

**Comment by S11001001** [comment]
> Compared to preceding an offset message before the actual query, this can start streaming immediately right after receiving the message, and we don't have to deal with timeout between offset and query

Neither of these are concerns in this case. In the first, there's no real difference for us in awaiting either one or two messages, regardless we start streaming immediately when we receive the query. In the second, there's less resource usage in awaiting the second message than streaming event blocks, so we can just use the overall stream timeout already in place.

To be completely clear, this _would never_ await an offset message _after_ the query message.

> or separate error handle per message.

This is a slight complication, though. On the other hand, probably only slight.



> I don't see this approach is a query format change.

By "query format change", I mean the request format for the WebSocket endpoints, not that of a different endpoint. With a preceding message, there is absolutely no change for existing use cases.

**Comment by lima-da** [comment]
@S11001001 
OK,
Will the initial offset(optional) message emit a response from json-api?
just a little worried about the disconnection between offset message and query here. and it also slightly complicate client implementation. because this introduce a "state" in the stream, instead of straight ask/response pattern.

**Comment by S11001001** [comment]
> Will the initial offset(optional) message emit a response from json-api?

No.

> and it also slightly complicate client implementation. because this introduce a "state" in the stream, instead of straight ask/response pattern.

There's no "state" for the client here; just send two messages, one after the other.

**Comment by lima-da** [comment]
ok, then I am fine with it if we can strictly define that only 
-> {offset: 1}
-> [query1, query2,...]
<- {"events": ...}
or 
-> [query1, query2,...]
<- {"events": ...}
would work.
